contexts/credentials/v2

```diff
@@ -112,26 +112,29 @@
       }
     },
 
-    "JsonSchema2023": {
+
+    "JsonSchemaCredential": {
       "@id":
-        "https://w3.org/2018/credentials#JsonSchema2023",
+        "https://w3.org/2018/credentials#JsonSchemaCredential",
       "@context": {
```

Comment:
Hmm -- this scoped context has now moved into `JsonSchemaCredential`. It should be under `JsonSchema` instead. That way, when an object of "type" `JsonSchema` is used, it will have the `jsonSchema` property (where the actual JSON schema goes) defined. It's that object that has that property, not the `JsonSchemaCredential`.

Reply:
@dlongley that seems revered to me. Perhaps examples make this clearer:

## Case 1
`JsonSchema` ... a type that refers to a JSON Object that validates JSON... (we intend to implement this one).

```json
{
  "@context": [
    "https://www.w3.org/ns/credentials/v2",
    "https://www.w3.org/ns/credentials/examples/v2"
  ],
  "id": "https://example.com/credentials/3732",
  "type": ["VerifiableCredential", "EmailCredential"],
  "issuer": "https://example.com/issuers/14",
  "issuanceDate": "2010-01-01T19:23:24Z",
  "credentialSubject": {
    "id": "did:example:ebfeb1f712ebc6f1c276e12ec21",
    "emailAddress": "subject@example.com"
  },
  "credentialSchema": {
    "id": "https://example.com/schemas/email.json",
    "type": "JsonSchema"
  }
}
```

## Case 2
`JsonSchemaCredential` ... a type that refers to a VerifiableCredential, that contains a json object, that validates JSON... (we do not plan to implement this one).

```json
{
  "@context": [
      "https://www.w3.org/ns/credentials/v2",
      "https://www.w3.org/ns/credentials/examples/v2"
  ],
  "id": "https://example.com/credentials/3734",
  "type": ["VerifiableCredential", "JsonSchemaCredential"],
  "issuer": "https://example.com/issuers/14",
  "issuanceDate": "2010-01-01T19:23:24Z",
  "credentialSubject": {
    "id": "https://example.com/schemas/email-credential-schema.json",
    "type": "JsonSchema",
    "jsonSchema": {
       "$id": "https://example.com/schemas/email-credential-schema.json",
       "$schema": "https://json-schema.org/draft/2020-12/schema",
       "name": "EmailCredential",
       "description": "EmailCredential using JsonSchemaCredential",
       "type": "object",
       "properties": {
         "credentialSubject": {
           "type": "object",
           "properties": {
             "emailAddress": {
               "type": "string",
               "format": "email"
             }
           },
           "required": ["emailAddress"]
         }
       }
    }
  }
}
```

Reply:
Perhaps you are suggesting this?

```json
{
  "@context": [
      "https://www.w3.org/ns/credentials/v2",
      "https://www.w3.org/ns/credentials/examples/v2"
  ],
  "id": "https://example.com/credentials/3734",
  "type": ["VerifiableCredential", "JsonSchemaCredential"], // type
  "issuer": "https://example.com/issuers/14",
  "issuanceDate": "2010-01-01T19:23:24Z",
  "credentialSubject": {
       "$id": "https://example.com/schemas/email-credential-schema.json",
       "$schema": "https://json-schema.org/draft/2020-12/schema",
       "name": "EmailCredential",
       "description": "EmailCredential using JsonSchemaCredential",
       "type": "object",
       "properties": {
         "credentialSubject": {
           "type": "object",
           "properties": {
             "emailAddress": {
               "type": "string",
               "format": "email"
             }
           },
           "required": ["emailAddress"]
         }
       }
    }
  }
}
```

Reply:
@OR13,

My suggestions will make Case 1 and Case 2 work. The "Perhaps you were suggesting this" is not something I was suggesting and don't believe we should do.

Reply:
@OR13,

Right now, this PR will create these term mappings:

```json
    "JsonSchemaCredential": {
      "@id":
        "https://w3.org/2018/credentials#JsonSchemaCredential",
      "@context": {
        "@protected": true,

        "id": "@id",
        "type": "@type",

       "jsonSchema": {
          "id": "https://w3.org/2018/credentials#JsonSchema",
          "type": "@json"
       }
    },

    "JsonSchema": {
      "@id": "https://w3.org/2018/credentials#JsonSchema",
      "@context": {
        "@protected": true,

        "id": "@id",
        "type": "@type"
    },
```

This would mean that `jsonSchema` would only be defined at the top-level of a VC that used the type `JsonSchemaCredential`. Something like this (which we do not want):

```json
{
  "@context": [
      "https://www.w3.org/ns/credentials/v2",
      "https://www.w3.org/ns/credentials/examples/v2"
  ],
  "id": "https://example.com/credentials/3734",
  "type": ["VerifiableCredential", "JsonSchemaCredential"], // type which
  // defines the scoped-context with "jsonSchema" for use on this object
  // which we do not want
  "issuer": "https://example.com/issuers/14",
  "issuanceDate": "2010-01-01T19:23:24Z",
  // we do not want this
  "jsonSchema": {
     "$id": "https://example.com/schemas/email-credential-schema.json",
     "$schema": "https://json-schema.org/draft/2020-12/schema",
     "name": "EmailCredential",
     "description": "EmailCredential using JsonSchemaCredential",
     "type": "object",
     "properties": {
       "credentialSubject": {
         "type": "object",
         "properties": {
           "emailAddress": {
             "type": "string",
             "format": "email"
           }
         },
         "required": ["emailAddress"]
       }
     }
  },
  "credentialSubject": {
    "id": "https://example.com/schemas/email-credential-schema.json",
    "type": "JsonSchema"
  }
}
```

Instead, it should do this:

```json
    "JsonSchemaCredential": "https://w3.org/2018/credentials#JsonSchemaCredential",

    "JsonSchema": {
      "@id": "https://w3.org/2018/credentials#JsonSchema",
      "@context": {
        "@protected": true,

        "id": "@id",
        "type": "@type",

       "jsonSchema": {
          "id": "https://w3.org/2018/credentials#JsonSchema",
          "type": "@json"
       }
    },
```

Which will enable Case 2:

```json
{
  "@context": [
      "https://www.w3.org/ns/credentials/v2",
      "https://www.w3.org/ns/credentials/examples/v2"
  ],
  "id": "https://example.com/credentials/3734",
  "type": ["VerifiableCredential", "JsonSchemaCredential"], // type
  "issuer": "https://example.com/issuers/14",
  "issuanceDate": "2010-01-01T19:23:24Z",
  "credentialSubject": {
    "id": "https://example.com/schemas/email-credential-schema.json",
    "type": "JsonSchema", // this type has the scoped-context that
    // defines "jsonSchema" for use on this object
    "jsonSchema": {
       "$id": "https://example.com/schemas/email-credential-schema.json",
       "$schema": "https://json-schema.org/draft/2020-12/schema",
       "name": "EmailCredential",
       "description": "EmailCredential using JsonSchemaCredential",
       "type": "object",
       "properties": {
         "credentialSubject": {
           "type": "object",
           "properties": {
             "emailAddress": {
               "type": "string",
               "format": "email"
             }
           },
           "required": ["emailAddress"]
         }
       }
    }
  }
}
```

It also enables Case 1 by simply defining `JsonSchema` as a usable type at all.

Reply:
Strong disagreement.... I think.

I don't want to make every instance of "JsonSchema" a verifiable credential.

I don't even want to use "JsonSchemaCredential" at all.

My goal is to just use JSON Schema with `credentialSchema` property.

All I need in the v2 context to achieve this is this:

```json
"JsonSchema": {
      "@id": "https://w3.org/2018/credentials#JsonSchema",
}
```
All I need in a credential to achieve this is:

```json
 "credentialSchema": {
    "id": "https://example.com/schemas/email-credential-schema.json",
    "type": "JsonSchema" // type
  }
```

Reply:
> 

huge +1 to this - pretty hard objection to this pr as it stands

Reply:
@dlongley if your suggestion makes case 1 and case 2 true then I'll update the PR

Reply:
@decentralgabe,

> @dlongley if your suggestion makes case 1 and case 2 true then I'll update the PR

Yup, that's what it does. You'll have to figure out whether @OR13 and @mprorock will take issue with you supporting both cases. They appear to only want to support Case 1, but I'm not convinced that there isn't some confusion here.

Reply:
@OR13,

> Strong disagreement.... I think.
> 
> I don't want to make every instance of "JsonSchema" a verifiable credential.

Me either. This doesn't do that.

> I don't even want to use "JsonSchemaCredential" at all.

Yes, but @decentralgabe wants to.

> My goal is to just use normal OpenJS Foundation spec conformant JSON Schemas with `credentialSchema` property.

That's fine with what I've proposed.

> All I need in the v2 context to achieve this is this:
> 
> ```json
> "JsonSchema": {
>       "@id": "https://w3.org/2018/credentials#JsonSchema",
> }
> ```

True. But if you add a scoped context, you get to do what you want and @decentralgabe gets to do what he wants too.

> All I need in a credential to achieve this is:
> 
> ```json
>  "credentialSchema": {
>     "id": "https://example.com/schemas/email-credential-schema.json",
>     "type": "JsonSchema" // type
>   }
> ```

True. You still get to do this with my suggestion.

> I agree regarding not wanting the "JsonSchemaCredential" and its dependence on the "jsonSchema" predicate... but please don't break "JsonSchema" when fixing that problem : )

I don't think something is breaking. What specifically do you think is breaking and how?

Reply:
@dlongley I've updated the PR to reflect your changes ( I think )

Reply:
@dlongley  one question -- I would assume since we want to expose the term `jsonSchema` inside the `credentialSubject` of a `JsonSchemaCredential` then the term would need to be defined under `JsonSchemaCredential` why is it that there's no need to define `jsonSchema` within the `JsonSchemaCredential` type?

Reply:
There's a situation I don't quite understand. Imagine the following scenario

The VC has the property below.
```json
{
  "credentialSchema": {"id": "https://example.com/my-json-schema.json", "type": "JsonSchema"}
}
```

And the contents that result from dereferencing `https://example.com/my-json-schema.json` are the following:
```json
    {
       "$id": "https://example.com/my-json-schema.json",
       "$schema": "https://json-schema.org/draft/2020-12/schema",
       "name": "EmailCredential",
       "type": "object",
       "properties": {}
    }
```

Would the `VC.credentialSchema.type` conflict with `dereferenceURL(VC.credentialSchema.id).type`? Note that the result of `dereferenceURL` does not contain any `jsonSchema` property within it.

Reply:
@decentralgabe,

> @dlongley one question -- I would assume since we want to expose the term `jsonSchema` inside the `credentialSubject` of a `JsonSchemaCredential` then the term would need to be defined under `JsonSchemaCredential` why is it that there's no need to define `jsonSchema` within the `JsonSchemaCredential` type?

Because `jsonSchema` is not a property of an object of type `JsonSchemaCredential`. It's a property of an object of type `JsonSchema`.

Type-scoped contexts are applied to the objects where the `type` is declared. It's similar to OOP where you declare the properties that can appear on an object of a certain type. An object of type `JsonSchema` can use the `jsonSchema` property. We want that property to be defined on an object of that type, not on `JsonSchemaCredential`.

As an example:

```
{
  "@context": {
    "pet": "example:pet",
    "id": "@id",
    "type": "@type",
    "Dog": {
      "@id": "example:Dog",
      "@context": {
        "id": "@id",
        "type": "@type",
        "name": "example:name",
        "breed": "example:breed",
        "favoriteToy": "example:favoriteToy"
        // etc.
      }
    }
  },
  "pet": {
    "id": "example:some_id",
    "type": "Dog",
    "name": "Bark Hamill",
    "breed": "Chihuahua"
    // "favoriteToy" you don't have to use every defined property
  }
}
```

Reply:
@andresuribe87,

> There's a situation I don't quite understand. Imagine the following scenario
> 
> The VC has the property below.
> 
> ```json
> {
>   "credentialSchema": {"id": "https://example.com/my-json-schema.json", "type": "JsonSchema"}
> }
> ```
> 
> And the contents that result from dereferencing `https://example.com/my-json-schema.json` are the following:
> 
> ```json
>     {
>        "$id": "https://example.com/my-json-schema.json",
>        "$schema": "https://json-schema.org/draft/2020-12/schema",
>        "name": "EmailCredential",
>        "type": "object",
>        "properties": {}
>     }
> ```
> 
> Would the `VC.credentialSchema.type` conflict with `dereferenceURL(VC.credentialSchema.id).type`? Note that the result of `dereferenceURL` does not contain any `jsonSchema` property within it.

Just because a term mapping exists in a JSON-LD context does not mean it has to be used. Also, since the content at that URL is not `application/ld+json`, there's no reason to think (nor any automated way) its contents should be merged into a knowledge graph. If you wanted to enable some kind of automated way to merge, you could also serve `application/ld+json` from that URL with a different representation (when requested via content negotiation), but that doesn't sound like it's really even a use case here.